rootston/layer_shell.c

```diff
@@ -0,0 +1,324 @@
+				box.width, box.height, box.x, box.y);
+		wlr_layer_surface_configure(layer, box.width, box.height);
+	}
+}
+
+static void arrange_layers(struct wlr_output *_output) {
+	struct roots_output *output = _output->data;
+
+	struct wlr_box usable_area = { 0 };
+	wlr_output_effective_resolution(output->wlr_output,
+			&usable_area.width, &usable_area.height);
+
+	// Arrange exclusive surfaces from top->bottom
+	arrange_layer(output->wlr_output,
+			&output->layers[ZWLR_LAYER_SHELL_V1_LAYER_OVERLAY],
+			&usable_area, true);
+	arrange_layer(output->wlr_output,
+			&output->layers[ZWLR_LAYER_SHELL_V1_LAYER_TOP],
+			&usable_area, true);
+	arrange_layer(output->wlr_output,
+			&output->layers[ZWLR_LAYER_SHELL_V1_LAYER_BOTTOM],
+			&usable_area, true);
+	arrange_layer(output->wlr_output,
+			&output->layers[ZWLR_LAYER_SHELL_V1_LAYER_BACKGROUND],
+			&usable_area, true);
+	memcpy(&output->usable_area, &usable_area, sizeof(struct wlr_box));
+
+	struct roots_view *view;
+	wl_list_for_each(view, &output->desktop->views, link) {
+		if (view->maximized) {
+			view_arrange_maximized(view);
+		}
+	}
+
+	// Arrange non-exlusive surfaces from top->bottom
+	wlr_output_effective_resolution(output->wlr_output,
+			&usable_area.width, &usable_area.height);
+	arrange_layer(output->wlr_output,
+			&output->layers[ZWLR_LAYER_SHELL_V1_LAYER_OVERLAY],
+			&usable_area, false);
+	arrange_layer(output->wlr_output,
+			&output->layers[ZWLR_LAYER_SHELL_V1_LAYER_TOP],
+			&usable_area, false);
+	arrange_layer(output->wlr_output,
+			&output->layers[ZWLR_LAYER_SHELL_V1_LAYER_BOTTOM],
+			&usable_area, false);
+	arrange_layer(output->wlr_output,
+			&output->layers[ZWLR_LAYER_SHELL_V1_LAYER_BACKGROUND],
+			&usable_area, false);
+}
+
+static void handle_output_destroy(struct wl_listener *listener, void *data) {
+	struct roots_layer_surface *layer =
+		wl_container_of(listener, layer, output_destroy);
+	layer->layer_surface->output = NULL;
+	wl_list_remove(&layer->output_destroy.link);
+	wl_list_remove(&layer->output_mode.link);
+	wlr_layer_surface_close(layer->layer_surface);
+}
+
+static void handle_output_mode(struct wl_listener *listener, void *data) {
+	arrange_layers((struct wlr_output *)data);
+}
+
+static void handle_surface_commit(struct wl_listener *listener, void *data) {
+	struct roots_layer_surface *layer =
+		wl_container_of(listener, layer, surface_commit);
+	struct wlr_layer_surface *layer_surface = layer->layer_surface;
+	struct wlr_output *wlr_output = layer_surface->output;
+	if (wlr_output != NULL) {
+		struct roots_output *output = wlr_output->data;
+		output_damage_from_local_surface(output, layer_surface->surface,
+				layer->geo.x, layer->geo.y, 0);
+		arrange_layers(wlr_output);
+		output_damage_from_local_surface(output, layer_surface->surface,
+				layer->geo.x, layer->geo.y, 0);
+	}
+}
+
+static void unmap(struct wlr_layer_surface *layer_surface) {
+	struct roots_layer_surface *layer = layer_surface->data;
+	if (layer->link.prev) {
```

Comment:
meh

Reply:
meh

Reply:
Try to remove this check. If it doesn't work, this should:

```c
wl_list_remove(&layer->link);
wl_list_init(&layer->link);
```